### PR TITLE
Stop Details Use Case

### PR DESCRIPTION
**Pull request by KuzeyOzturac**

### CODE REVIEW

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsController.java
Comment (prash-red) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsController.java:1]: This file should be moved to interface_adapters.

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsService.java
Comment (prash-red) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsService.java:1]: The service file should be an interface between the controller and the implementation

**Review by prash-red — CHANGES_REQUESTED** [review]
The overall code looks good. Make an implementation file that implements the actual use-case. Service is supposed to be an interface between the controller and implementation. Look at how I implemented the show_stops use-case.

**Review by prash-red — COMMENTED** [review]
Looks good !👌

File: backend/src/main/java/com/example/backend/interface_adapters/StopDetailsController.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+@CrossOrigin(origins = "*")
+public class StopDetailsController {
+    /**
+    *  Instance variables:
+    *  - stopDetailsService: The service for this controller
+    */
+    @Autowired
+    private StopDetailsService stopDetailsImplementation;
+
+//    public StopDetailsController(StopDetailsService stopDetailsService){
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/interface_adapters/StopDetailsController.java:23]: Remove this comment

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsImplementation.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+     * Instance variables:
+     *  - routeDataAccessObject: the data access object used to access data about routes (such as routeTags)
+     *  - stopDataAccessObject: the data access object used to access data about stops (such as stopName)
+     */
+    @Autowired
+    private RouteDataAccessInterface routeDataAccessObject;
+    @Autowired
+    private StopDataAccessInterface stopDataAccessObject;
+
+//    public StopDetailsService(StopDataAccessInterface stopDataAccessObject, RouteDataAccessInterface routeDataAccessObject){
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsImplementation.java:31]: Remove this comment

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsInputData.java
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsInputData.java:1]: Why not pass this as a parameter?

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsOutputData.java [hunk 3]
```diff
@@ -1,7 +1,34 @@
...
 public class StopDetailsOutputData {
-
-    public StopDetailsOutputData() {
+    /**
+     * Instance variables:
+     * - stopName: the name of the stop
+     * - routeTagsToWrapper: maps route tags to route name, and direction tags mapped to direction names
+     */
+    private String stopName;
+    private HashMap<String, ArrayList<Object>> routeTagsToWrapper;
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsOutputData.java:16]: Return a mapping of Route Tag to a mapping of direction tag to direction name.

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsImplementation.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+//        this.stopDataAccessObject = stopDataAccessObject;
+//        this.routeDataAccessObject = routeDataAccessObject;
+//    }
+
+    /**
+     * Given an input data, extracts the stopTag and fetches the name of the corresponding stop.
+     * Finds the route tags of all the routes that pass through the stop and puts them into a set called routeTags.
+     * Creates the output data containing the stopName and the routeTags, and returns it.
+     */
+    public StopDetailsOutputData execute(StopDetailsInputData inputData){
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsImplementation.java:41]: Update implementation to reflect changes in the OutputData.

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsOutputData.java [hunk 5]
```diff
@@ -1,7 +1,34 @@
...
+     */
+    public StopDetailsOutputData(String stopName, HashMap<String, ArrayList<Object>> routeTagsToWrapper) {
+        this.stopName = stopName;
+        this.routeTagsToWrapper = routeTagsToWrapper;
     }
-}
+    /**
+    * Returns the stopName
+    */
+    public String getName(){return this.stopName;}
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsOutputData.java:27]: Rename to getStopName()

File: backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import com.example.backend.data_access.stop.StopDataAccessInterface;
+
+/**
+ * This is the service interface for the stop details use case.
+ */
+
+public interface StopDetailsService {
+
+    /**
+     * Given an input data, extracts the stopTag and fetches the name of the corresponding stop.
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/use_case/stop_details/StopDetailsService.java:13]: Remove comment

**Review by nsaroiu — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/example/backend/interface_adapters/StopDetailsController.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    *  Instance variables:
+    *  - stopDetailsService: The service for this controller
+    */
+    @Autowired
+    private StopDetailsService stopDetailsImplementation;
+
+    /**
+     * Given a stopTag, creates the input data and executes the service with it.
+     */
+    @PostMapping("/stopdetails")
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/interface_adapters/StopDetailsController.java:23]: change endpoint to "/stop-details"

**Review by nsaroiu — CHANGES_REQUESTED** [review]
Just make sure to change the endpoint, otherwise it is perfect.

**Review by nsaroiu — APPROVED** [review]
Looking good, changes address all feedback

**Review by prash-red — APPROVED** [review]
Changes address all feedback given.

**Pull request merged**